[quant] make quantize_per_tensor/per_channel accept qscheme argument

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70324
* #70323

Summary:
qscheme is used to denote which mode was used to calculate the scale/zero_point for the tensor

Test Plan:
```
>>> x = torch.rand(2, 2)
>>> qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8, torch.per_tensor_affine)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_tensor_affine, scale=1.0, zero_point=0)
>>> qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8, torch.per_tensor_symmetric)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_tensor_symmetric, scale=1.0, zero_point=0)

>>> qx = torch.quantize_per_channel(x, torch.tensor([1.0, 1.0]), torch.tensor([0, 0]), 0, torch.quint8, torch.per_channel_symmetric)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_channel_symmetric,
       scale=tensor([1., 1.], dtype=torch.float64), zero_point=tensor([0, 0]),
       axis=0)
>>> qx = torch.quantize_per_channel(x, torch.tensor([1.0, 1.0]), torch.tensor([0, 0]), 0, torch.quint8, torch.per_channel_affine)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_channel_affine,
       scale=tensor([1., 1.], dtype=torch.float64), zero_point=tensor([0, 0]),
       axis=0)
```

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
[quant] make quantize_per_tensor/per_channel accept qscheme argument
Summary:
qscheme is used to denote which mode was used to calculate the scale/zero_point for the tensor

Test Plan:
```
>>> x = torch.rand(2, 2)
>>> qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8, torch.per_tensor_affine)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_tensor_affine, scale=1.0, zero_point=0)
>>> qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8, torch.per_tensor_symmetric)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_tensor_symmetric, scale=1.0, zero_point=0)

>>> qx = torch.quantize_per_channel(x, torch.tensor([1.0, 1.0]), torch.tensor([0, 0]), 0, torch.quint8, torch.per_channel_symmetric)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_channel_symmetric,
       scale=tensor([1., 1.], dtype=torch.float64), zero_point=tensor([0, 0]),
       axis=0)
>>> qx = torch.quantize_per_channel(x, torch.tensor([1.0, 1.0]), torch.tensor([0, 0]), 0, torch.quint8, torch.per_channel_affine)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_channel_affine,
       scale=tensor([1., 1.], dtype=torch.float64), zero_point=tensor([0, 0]),
       axis=0)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -112,31 +133,31 @@ std::vector<Tensor> dequantize_tensors_quantized_cpu(TensorList tensors) {
 double q_scale_quant(const Tensor& self) {
   auto quantizer = get_qtensorimpl(self)->quantizer();
-  TORCH_CHECK(quantizer->qscheme() == kPerTensorAffine);
+  TORCH_CHECK(quantizer->qscheme() == kPerTensorAffine || quantizer->qscheme() == kPerTensorSymmetric);
```
nit: maybe we can create a helper function for this check

jerryzh168 commented on aten/src/ATen/native/quantized/QTensor.cpp:
```diff
@@ -93,6 +103,17 @@ Tensor quantize_per_channel(
   return quantizer->quantize(self);
 }
 
+Tensor quantize_per_channel_w_qscheme(
```
what about adding a qscheme with default argument of per_tensor_affine/per_channel_affine to quantize_per_tensor/quantize_per_channel functions?

supriyar replied:
Since QScheme is a special type, we cannot directly assign a default value to it in `native_functions.yaml` for existing `quantize_per_tensor` function. Let me know if you're aware of a way to do this.

jerryzh168 replied:
is this a limitation of codegen?

jerryzh168 replied:
I think we have some defaults for QScheme, like https://github.com/pytorch/pytorch/blob/master/tools/codegen/api/python.py#L283, can you try using `per_tensor_affine` in `native_functions.yaml`

supriyar replied:
I don't think the defaults for qscheme are supported in the codegen for `native_functions.yaml`. If we add QScheme argument to `quantize_per_tensor` it throws the following error https://gist.github.com/supriyar/0165a7013761ea1dfd5118d2ee7466ca

supriyar committed:
Update on "[quant] make quantize_per_tensor/per_channel accept qscheme argument"


Summary:
qscheme is used to denote which mode was used to calculate the scale/zero_point for the tensor

Test Plan:
```
>>> x = torch.rand(2, 2)
>>> qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8, torch.per_tensor_affine)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_tensor_affine, scale=1.0, zero_point=0)
>>> qx = torch.quantize_per_tensor(x, 1.0, 0, torch.quint8, torch.per_tensor_symmetric)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_tensor_symmetric, scale=1.0, zero_point=0)

>>> qx = torch.quantize_per_channel(x, torch.tensor([1.0, 1.0]), torch.tensor([0, 0]), 0, torch.quint8, torch.per_channel_symmetric)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_channel_symmetric,
       scale=tensor([1., 1.], dtype=torch.float64), zero_point=tensor([0, 0]),
       axis=0)
>>> qx = torch.quantize_per_channel(x, torch.tensor([1.0, 1.0]), torch.tensor([0, 0]), 0, torch.quint8, torch.per_channel_affine)
>>> qx
tensor([[0., 0.],
        [0., 1.]], size=(2, 2), dtype=torch.quint8,
       quantization_scheme=torch.per_channel_affine,
       scale=tensor([1., 1.], dtype=torch.float64), zero_point=tensor([0, 0]),
       axis=0)
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]
